Write node color space

lucienfostier opened the pull request:
This introduces color management for the ImageWriter.
-----------

Allow to register a function to set a default colorspace based on ImageWriter plugs value or override the colorspace for the output image.

- ImageWriter:
 - Added ColorSpace plug to use default colorspace or user override.
 - Support for config based default colorspace.

lucienfostier commented:
@johnhaddon,

based on our discussions, I've backed up regarding storing the colorspace of output image into metadata.
I think this can be done other way if the user really wants to do it.

I've left the commit structure a bit messy for now so we can discuss the change before I do some clean up.

I think I have still some works to do regarding the test but at least we can start with this.

johnhaddon commented on include/GafferImage/ImageWriter.h:
```diff
@@ -37,6 +37,8 @@
 #ifndef GAFFERIMAGE_IMAGEWRITER_H
 #define GAFFERIMAGE_IMAGEWRITER_H
 
+#include "Gaffer/Expression.h"
```
To keep the headers as clean as we can, let's use a forward declaration (in the section starting at line 46 below) rather than include Expression.h here.

johnhaddon commented on src/GafferImage/ImageWriter.cpp:
```diff
@@ -706,10 +711,24 @@ ImageWriter::ImageWriter( const std::string &name )
+
+	ExpressionPtr expression = new Expression( "__expression" );
+	addChild( expression );
+	//expression->setExpression( string( "parent[\"__colorSpace\"][\"outputSpace\"] = context.get(\"gaffer:colorspace\", \"\") " ), string( "python" ) );
```
Remove dead code.

johnhaddon commented on src/GafferImage/ImageWriter.cpp:
```diff
@@ -706,10 +711,24 @@ ImageWriter::ImageWriter( const std::string &name )
 	addChild( new StringPlug( "fileName" ) );
 	addChild( new StringPlug( "channels", Gaffer::Plug::In, "*" ) );
+
+	ColorSpacePtr colorSpace = new ColorSpace( "__colorSpace" );
```
I wonder, will it be necessary to promote the `ColorSpace::contextPlug()` here for some workflows?

lucienfostier replied:
I guess you could to be flexible but in practice I think this is unlikely to be useful.
ColorSpaces used in stored files are very unlikely to use context dependent ones.
Let me know if you still think it's a good idea to do so, I think it should be trivial.

johnhaddon replied:
No need to do it now if you don't think it'd be useful. As you say, it'll be trivial to add later if necessary.

johnhaddon commented on src/GafferImage/ImageWriter.cpp:
```diff
@@ -706,10 +711,24 @@ ImageWriter::ImageWriter( const std::string &name )
+	addChild( expression );
+	//expression->setExpression( string( "parent[\"__colorSpace\"][\"outputSpace\"] = context.get(\"gaffer:colorspace\", \"\") " ), string( "python" ) );
+
+	addChild( new StringPlug( "colorSpace" ) );
```
This is a minor thing, but generally I've found it more readable to add all the public plugs _first_, at the top of the function, and then all the implementation details after (in which case this would move to line 714). That gets you into a "here's my interface, and here's my implementation" mindset, and also means the order of the plug/node accessors in the .cpp file matches that in the header.

johnhaddon commented on src/GafferImage/ImageWriter.cpp:
```diff
@@ -852,6 +913,11 @@ IECore::MurmurHash ImageWriter::hash( const Context *context ) const
 	h.append( channelsPlug()->hash() );
+	if( colorSpacePlug()->getValue() == "")
+	{
+		cout<<"append default color space to hash: "<<defaultColorSpace()( currentFileFormat() )<<endl;
```
Remove debug output.

johnhaddon replied:
I think we also need to hash in the value when it's `!= ""`?

johnhaddon commented on src/GafferImage/ImageWriter.cpp:
```diff
@@ -869,6 +935,12 @@ IECore::MurmurHash ImageWriter::hash( const Context *context ) const
 void ImageWriter::execute() const
 {
 	// Create an OIIO::ImageOutput
+	Context * c = const_cast<Context *>( Context::current() );
```
The context is const for a very good reason, so we should make a copy before editing it. I have a PR up that I hope will be merged today, which will make that even easier, looking like this :

```
Context::EditableScope colorSpaceScope( Context::current() );
colorSpaceScope.set( "gaffer:colorspace", defaultColorSpace()( currentFileFormat() );
```

johnhaddon commented on src/GafferImage/ImageWriter.cpp:
```diff
@@ -869,6 +935,12 @@ IECore::MurmurHash ImageWriter::hash( const Context *context ) const
+	c->set( "gaffer:colorspace", defaultColorSpace()( currentFileFormat() ) );
+
+
+	Expression *e = const_cast<Expression *>( expression() );
```
Setting the expression is classed as a graph edit, and it's forbidden to make those from within `execute()` (because threaded execution could mean that another call to `execute()` is already in flight). From our previous discussions I know why we can't put this in the constructor where we'd like it though, so I'm not 100% sure what to do. One thought would be to make a very simple ComputeNode to return the value from the context, so we don't need the Expression node at all. Although that's a bit more code, it would be nice to remove the dependency on Expression (and Python) from ImageWriter. Don't worry about this immediately, but I think I might take a look at it before we merge, or propose that we do it soon after.

lucienfostier replied:
Alright, sounds good!
When we discussed this topic initially with Daniel and Ivan that was one of the suggestion indeed.
Could this ComputeNode even works without the context at all, based on input plug use the registered function to compute a colorspace name and set an outPlug with that value?

Another thing I'm wondering is when you say remove the dependency on Python, you mean only the execution of the expression because correct me if I'm wrong but computing the colorSpace name still goes through python but I assume this is cached? Does the Expression using the Context also cache efficiently?

johnhaddon replied:
> Alright, sounds good!

Cool. Let's do that then.

> Could this ComputeNode even works without the context at all, based on input plug use the registered function to compute a colorspace name and set an outPlug with that value?

The problem would be that the ComputeNode would need access to all the plugs of the ImageWriter, but it only has access to its own plugs. In some ways it'd be nice if the ImageWriter was a ComputeNode as well as a TaskNode, but I don't think that's a good idea generally.

> Another thing I'm wondering is when you say remove the dependency on Python, you mean only the execution of the expression because correct me if I'm wrong but computing the colorSpace name still goes through python but I assume this is cached?

Correct. I'm suggesting we remove the dependency on Python at the _library_ level, which would mean it was still possible to use ImageWriter from a pure C++ application linked to libGafferImage. At the _application_ level, we still have the python dependency because the application is written in Python and the config uses python, but that's fine, that's how the main Gaffer app is written. We just want to keep the library clean so it could be used independently if necessary.

> Does the Expression using the Context also cache efficiently?

Yes. The hash for the python expression (which forms the key for the cache lookup) is always computed entirely in C++, this is what allows python expressions to outperform python nodes. In this case the hash will be composed only from the context variable, so we should execute the expression only once for each unique value of the variable.

johnhaddon commented on src/GafferImageBindings/ImageWriterBinding.cpp:
```diff
@@ -42,16 +44,48 @@
 using namespace GafferImage;
 using namespace GafferDispatchBindings;
 
+struct DefaultColorSpaceFunctionCreator
```
Should be inside an anonymous namespace. Also, the "Creator" part of the name seems a bit misleading, maybe just `DefaultColorSpaceFunction`?

johnhaddon commented on src/GafferImageBindings/ImageWriterBinding.cpp:
```diff
@@ -42,16 +44,48 @@
+		object m_fn;
+};
+
+static void registerDefaultColorSpace( object f )
```
Once this is inside the anonymous namespace, you can remove the `static` prefix.

johnhaddon commented on src/GafferImage/ImageWriter.cpp:
```diff
@@ -976,8 +975,11 @@ void ImageWriter::execute() const
+	const std::string sceneLinear = config->getColorSpace( OpenColorIO::ROLE_SCENE_LINEAR )->getName();
 	Expression *e = const_cast<Expression *>( expression() );
-	e->setExpression( string( "parent[\"__colorSpace\"][\"outputSpace\"] = context.get(\"gaffer:colorspace\", \"\") " ), string( "python" ) );
+	e->setExpression( boost::str( boost::format( "parent[\"__colorSpace\"][\"outputSpace\"] = context.get(\"gaffer:colorspace\", \"%s\") " )  % sceneLinear ), string( "python" )  );
```
Quick question - why do we need the default value at all? Shouldn't the expression only ever be computed in a context that contains the colorspace variable?

lucienfostier replied:
Ok I was just thinking it's good practice but I'll remove it if you think it's useless.

johnhaddon replied:
It's good practice for user expressions, because they have less control over the context the expression will be computed in. But in this case, we _know_ the variable must exist, so the default value would only hide any bugs that meant the variable didn't exist.

johnhaddon commented on src/GafferImage/ImageWriter.cpp:
```diff
@@ -869,6 +935,12 @@ IECore::MurmurHash ImageWriter::hash( const Context *context ) const
 {
 	// Create an OIIO::ImageOutput
+	Context * c = const_cast<Context *>( Context::current() );
+	c->set( "gaffer:colorspace", defaultColorSpace()( currentFileFormat() ) );
```
Let's call this something like  "__imageWriter:colorSpace", to indicate that it's a private implementation detail. At some point it might even be nice if we had a DeleteContextVariables node we could use to prevent it being visible upstream at all, but we can leave that for another time.

lucienfostier replied:
Are you talking a future node (DeleteContextVariables) that we dont have yet ?

danieldresser replied:
Yeah, he's talking about a potential future node.  It would be quite useful - there are places in my current work where I have to set loop indices to 0 instead of deleting them, which means that I'm unnecessarily hashing things twice ( I'd be hashing them a lot more than that if I didn't reset the index ).

So, yeah, DeleteContextVariables sounds like a great idea.

johnhaddon replied:
Cool - shouldn't take long for me to throw a DeleteContextVariables node together. Let's aim to get that done before we merge this.

lucienfostier commented:
Hey John,

I've addressed most of your comments, and added a few remarks or questions along the line.
I've also cleaned up the commit list.

I would like to talk about the test as well because after having a first look there might be some important change to be made.

First thing I realised is that expectation have changing regarding the colorspace used on output of the ImageWriter and on the input of the ImageReader.
For example, dpx test now will have different colorSpace based on their bitDepth.
One possible approach to solve this issue is to have a public method on the imageWriter accessible from python to query the colorspace and use that to set the ImageReader appropriately.

Besides, there are a bunch of test failing for a reason that eludes me at the moment (testCropDataWindowToDisplayWindowScanline).

Please let me know what are the next step to get this PR merged.

Cheers

johnhaddon commented:
> I would like to talk about the test as well because after having a first look there might be some important change to be made.

Good point! I should have thought of this earlier. I think we need the ImageReader and the ImageWriter to use the same colourspace config function, so that we can round-trip data between them reliably.

lucienfostier commented:
oops, I somehow failed to see your latest comment.

Anyway I pushed a couple of commit to go into that direction.

I took a stab at a node to fetch the context value, it's probably not as generic as you might wish but it was a good experience for me to try adding a new node from scratch.I'm happy to get that to the finish line under your guidance.

I also made a binding to get the default colorspace from the writer, I dont know if this is what you have in mind for the round trip...?Let me know if not.

Next I will have a look at the test again to see if I have everything I need to make them to work again.

I also need a rebase but the conflicts are simple.

johnhaddon commented:
Hi Lucien,

Thanks for the update. I'm afraid I owe you an apology. I've only just realised that the solution to our "get the context into the colourspace plug" is even simpler than an expression or a ContextValue node. Because `ColorSpace::outputSpacePlug()` is a StringPlug, we can simply put `"${__imageWriter:colorSpace}"` in it and it'll be substituted for us automatically. Sorry. I hope the experience writing a node hasn't been completely wasted.

For the round trip, I was thinking we'd expose the `DefaultColorSpaceFunction` so that the reader could call it directly - because the ImageReader it will need to use a different mechanism for getting and passing the filename and data type. Does that make sense?

Cheers...
John

lucienfostier commented:
Good point! I'll make the necessary adjustment.

I was thinking of binding 'defaultColorSpace' rather than 'defaultColorSpaceFunction', I'm replicating how it's called in the writer inside the reader.

I'm currently looking at that and one problem I stumbled upon is how to determined the bit depth of the input image, my hope was that the metadata "oiio:BitsPerSample" was define but it's not and that probably make sense. Another idea I wanted to explore is using the imageSpec class but I don't know if there is a prohibitive cost to that.If you have any suggestion regarding that, I'm all hear.

Besides, I would come back to just submitting the file format and the bit depth and possibly the full file path ( for asset management purpose) to the config file instead of the entire optionPlug values from the ImageWrite.The rationale being, I dont see how tiling/scanline or compression would influence colorspace.

Let me know what you think of all that.

cheers

johnhaddon commented:
> Good point! I'll make the necessary adjustment.

Great - thanks for your patience.

> I was thinking of binding 'defaultColorSpace' rather than 'defaultColorSpaceFunction', I'm replicating how it's called in the writer inside the reader.

I think I was confused because `defaultColorSpace()` doesn't take any arguments. But I presume you're thinking of making it static and adding arguments so that the ImageReader can pass the filename/bitdepth? In that case, that sounds great...

> I'm currently looking at that and one problem I stumbled upon is how to determined the bit depth of the input image, my hope was that the metadata "oiio:BitsPerSample" was define but it's not and that probably make sense. Another idea I wanted to explore is using the imageSpec class but I don't know if there is a prohibitive cost to that.If you have any suggestion regarding that, I'm all hear.

Using the ImageSpec sounds good to me - we [use a cache](https://github.com/GafferHQ/gaffer/blob/master/src/GafferImage/OpenImageIOReader.cpp#L104) to get access to those internally in the OpenImageIOReader, so we could use that. Potentially we could either add a new output plug to the OpenImageIOReader to pass the bit depth out, or we could automatically add "oiio:BitsPerSample" to the metadata in OpenImageIOReader. I suppose there are potential issues where a file format allows different bit depths per channel, but for our purposes it might be sufficient?

One things that might have a bearing on this is that we [force OIIO's cache to store floats](https://github.com/GafferHQ/gaffer/blob/master/src/GafferImage/OpenImageIOReader.cpp#L92). This might mean we get the wrong bit depth reported, but I think we can remove that float workaround now anyway probably.

>Besides, I would come back to just submitting the file format and the bit depth and possibly the full file path ( for asset management purpose) to the config file instead of the entire optionPlug values from the ImageWrite.The rationale being, I dont see how tiling/scanline or compression would influence colorspace.

Yep, sounds good.

Thanks Lucien!
Cheers...
John

lucienfostier commented:
Hey John,

I do not plan to add argument to defaultColorSpace, instead it wraps defaultColorSpaceFunction calls and compute the file format and bit depth in c++ for both image reader and writer.Is there a misunderstanding on this point between us, or am I getting this wrong?

Can you clarify the part about file format supporting multiple bit depth, I don't think it's a problem because I'm assuming the imageSpec return the native format of the pixel store in the files and not the pixel data loaded in Gaffer, correct?

Cheers

I already started implemented the out plug methodlogy, I'm planning to have a computeDataType called in the compute.I already verified that imageSpec->format gives me exactly what I need.

It doesnt seem that this force float affect the spec->format so I'm guessing we are good!

johnhaddon commented:
> I do not plan to add argument to defaultColorSpace, instead it wraps defaultColorSpaceFunction calls and compute the file format and bit depth in c++ for both image reader and writer.Is there a misunderstanding on this point between us, or am I getting this wrong?

There's definitely some misunderstanding, but I'm not sure where. I thought you had made the `ImageWriter::defaultColorSpace()` method public thinking that you would be able to use it from the ImageReader. But it only works for the ImageWriter because it is a method, and therefore implicitly has access to the settings on the ImageWriter. To get at the settings on an ImageReader, it would need to be passed an argument of some sort. I was expecting to pass filename/bitdepth but it sounds like maybe you are thinking of passing an `ImageReader *`? I think filename/bitdepth would be better, because it keeps the ImageReader internals private, and also allows other queries or the default colourspace (maybe it would be useful for a pipeline script to query it for instance).

> Can you clarify the part about file format supporting multiple bit depth, I don't think it's a problem because I'm assuming the imageSpec return the native format of the pixel store in the files and not the pixel data loaded in Gaffer, correct?

In a single EXR file for instance, I believe some channels could be float, some int, and some half. So there would be multiple bitdepths in a single file. But I think this is rare enough that we can ignore it (and EXRs will typically be linear anyway).

> I already started implemented the out plug methodlogy, I'm planning to have a computeDataType called in the compute.I already verified that imageSpec->format gives me exactly what I need.

Just to check we're on the same page, this is in the OpenImageIOReader, correct?

lucienfostier commented:
Ok I understand the misunderstanding a bit better.

I now have a smiliar api on the image reader and I'm registering the config for both node.

Ok got it for the multiple bit depth per file but I think it's an edge case and not part of best practice if you store different colorspace under the same file( exception would be render pass with raw data like pref or normal).

Yes its definitely on the OpenImageIOReader that I'm adding the plug sorry for the confusion.

lucienfostier commented:
I updated the expected dpx and tif because they were written in linear instead of cineon for one and srgb for the other.

lucienfostier commented:
I dont know how to interpret the tests.

Some have stalled like they do locally but on GafferImageTest.ImagePlugTest testImageHash, but I don't think I changed anything about that.

Some failed on Gaffer Test Failed : GafferRenderManTest but succeeded the  GafferImageTest.ImagePlugTest testImageHash...

Can you help me figure this one out?
I'd like to make sure current test pass before writing new ones and cleaning my commit before review.

Cheers

johnhaddon commented:
> Some have stalled like they do locally but on GafferImageTest.ImagePlugTest testImageHash, but I don't think I changed anything about that. Can you help me figure this one out?

I don't think the hangs are a problem with your code - I think you've just triggered a GIL management  bug that existed already. Could you cherry pick this fix I've just made and check that it works for you?

John

lucienfostier commented:
Hey John,

your patch seems to have fix the ImagePlug problem, thanks!

What do you reckon regarding the renderman test?

johnhaddon commented:
> your patch seems to have fix the ImagePlug problem, thanks!

Great!

> What do you reckon regarding the renderman test?

Do you mean this one?

```
FAIL: testWedge (GafferRenderManTest.RenderManRenderTest.RenderManRenderTest)
```

It looks like we're writing a tiff file in that test and then loading it with the ImageReader, so it seems likely that we're applying an sRGB->linear transform on loading and that is causing the failure? If so, I think we can just adjust the threshold in the test to account for the new behaviour...

johnhaddon commented:
Looks like we're getting there Lucien - could you squash into a smaller set of commits avoiding the ContextValue dead end etc?

johnhaddon commented:
Just remembered - we'll also need to run the GafferArnold tests manually since those don't run on Travis, and might be sensitive to colourspace changes too.

lucienfostier commented:
great!

I have one last commit to add before I squash to add test for the new features.

sure I'll run the arnold's one.

cheers

lucienfostier commented:
ok I have added tests for the new feature and run the arnold test locally.
Finally I cleaned up the commit, so feel free to have a look.

Cheers

johnhaddon commented on python/GafferImageUI/OpenImageIOReaderUI.py:
```diff
@@ -122,6 +122,16 @@
 
 		],
 
+		"__dataType" : [
```
Since this is a public plug (intended for external use), it doesn't need the "__" prefix.

lucienfostier replied:
sure!

johnhaddon commented on src/GafferImage/ImageWriter.cpp:
```diff
@@ -852,6 +922,15 @@ IECore::MurmurHash ImageWriter::hash( const Context *context ) const
+	}
+	else
+	{
+		h.append( colorSpacePlug()->getValue() );
```
Best to avoid a second call to `colorSpacePlug()->getValue()`, since hash functions are meant to be as past as possible. In fact, maybe we should replace the `defaultColorSpace()` method with a `colorSpace()`, since everywhere we call it we seem to be doing this check for an override anyway.

lucienfostier replied:
sounds good

johnhaddon commented on startup/GafferImage/defaultColorSpace.py:
```diff
@@ -0,0 +1,70 @@
+			"uint8"  : display,
+			"uint16" : display,
+			"float"  : sceneLinear,	#writing
+			"uint32"  : sceneLinear,	#reading
```
Seems to be some misalignment of the comments here. Use spaces rather than tabs to align (and tabs to indent only).

lucienfostier replied:
sure!

lucienfostier commented on src/GafferImage/OpenImageIOReader.cpp:
```diff
@@ -517,6 +533,25 @@ void OpenImageIOReader::hashFileName( const Gaffer::Context *context, IECore::Mu
+	const int bitsPerSample = spec->get_int_attribute( "oiio:BitsPerSample", 0 );
+	if( bitsPerSample )
+	{
+		return boost::str( boost::format( "uint%d"  ) % bitsPerSample );
```
https://github.com/OpenImageIO/oiio/blob/5b25ba0d36e6031831b7778b6c36423e68f7e8f5/src/tiff.imageio/tiffoutput.cpp#L97

seems to be ok?

lucienfostier replied:
I havent found any clues of this being signed in the OIIO repo.

lucienfostier commented on include/GafferImage/ImageReader.h:
```diff
@@ -110,10 +110,18 @@ class ImageReader : public ImageNode
 		Gaffer::StringPlug *colorSpacePlug();
 		const Gaffer::StringPlug *colorSpacePlug() const;
 
+		Gaffer::StringPlug *colorSpaceMetadataNamePlug();
```
Yes sure,

so this is useful if you have are storing different colorspaces using the same file format/bit depth for the same job.Example is show using wide gamut AcesCG and "standard" linear srgb primaries.
You can use the metadata to differentiate between what otherwise would have the exact same specs.

johnhaddon replied:
The thing I liked about the design we discussed is that it enforced some symmetry, so that we could have reasonable confidence that anything we wrote could be read back and would be the same. We had a single mechanism for automatic management of the colourspace (the function) that could be set up at a supervisor level and a single "i know best" mechanism (the colorSpace plug) that could be overridden at an artist level, but this seems to introduce a new mechanism that isn't quite either, and doesn't have equivalent support on the writer.

Maybe a compromise would be to provide the full image metadata to both the reader's and the writer's DefaultColorSpaceFunction, so that if some facilities/shows wanted to use metadata to control colourspace they could do so without putting it under user control?

lucienfostier replied:
alright, I think that would work with passing the full image metadata.

Could you provide an example to add the content of CompoundObjectPlug to the CompoundData, I'm assuming that I would use the ImagePlug->metadataPlug() as a source?

johnhaddon replied:
Yes, I think you would use the OpenImageIOReader.out.metadata plug. You could just call `getValue()` on it and then add it to your CompoundData I think :

```
compoundData->writable()["metadata"] = imagePlug->metadataPlug()->getValue();
```

lucienfostier replied:
ah!

`faut pas chercher midi a quatorze heure` as we say in french.

thanks

lucienfostier replied:
`In file included from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/intrusive_ptr.hpp:16:0,
                 from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/IECore/RefCounted.h:44,
                 from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/IECore/RunTimeTyped.h:39,
                 from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/IECore/Object.h:44,
                 from include/Gaffer/ValuePlug.h:41,
                 from include/Gaffer/StringPlug.h:41,
                 from src/GafferImage/ImageReader.cpp:41:
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp: In instantiation of 'boost::intrusive_ptr<T>& boost::intrusive_ptr<T>::operator=(const boost::intrusive_ptr<U>&) [with U = const IECore::CompoundObject; T = IECore::Data]':
src/GafferImage/ImageReader.cpp:281:31:   required from here
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:104:9: error: no matching function for call to 'boost::intrusive_ptr<IECore::Data>::intrusive_ptr(const boost::intrusive_ptr<const IECore::CompoundObject>&)'
         this_type(rhs).swap(*this);
         ^
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:104:9: note: candidates are:
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:90:5: note: boost::intrusive_ptr<T>::intrusive_ptr(const boost::intrusive_ptr<T>&) [with T = IECore::Data]
     intrusive_ptr(intrusive_ptr const & rhs): px( rhs.px )
     ^
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:90:5: note:   no known conversion for argument 1 from 'const boost::intrusive_ptr<const IECore::CompoundObject>' to 'const boost::intrusive_ptr<IECore::Data>&'
In file included from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/intrusive_ptr.hpp:16:0,
                 from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/IECore/RefCounted.h:44,
                 from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/IECore/RunTimeTyped.h:39,
                 from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/IECore/Object.h:44,
                 from include/Gaffer/ValuePlug.h:41,
                 from include/Gaffer/StringPlug.h:41,
                 from src/GafferImage/ImageReader.cpp:41:
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:76:5: note: template<class U> boost::intrusive_ptr<T>::intrusive_ptr(const boost::intrusive_ptr<U>&, typename boost::detail::sp_enable_if_convertible<Y, T>::type)
     intrusive_ptr( intrusive_ptr<U> const & rhs, typename boost::detail::sp_enable_if_convertible<U,T>::type = boost::detail::sp_empty() )
     ^
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:76:5: note:   template argument deduction/substitution failed:
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp: In substitution of 'template<class U> boost::intrusive_ptr<T>::intrusive_ptr(const boost::intrusive_ptr<U>&, typename boost::detail::sp_enable_if_convertible<Y, T>::type) [with U = const IECore::CompoundObject]':
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:104:9:   required from 'boost::intrusive_ptr<T>& boost::intrusive_ptr<T>::operator=(const boost::intrusive_ptr<U>&) [with U = const IECore::CompoundObject; T = IECore::Data]'
src/GafferImage/ImageReader.cpp:281:31:   required from here
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:76:5: error: no type named 'type' in 'struct boost::detail::sp_enable_if_convertible<const IECore::CompoundObject, IECore::Data>'
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp: In instantiation of 'boost::intrusive_ptr<T>& boost::intrusive_ptr<T>::operator=(const boost::intrusive_ptr<U>&) [with U = const IECore::CompoundObject; T = IECore::Data]':
src/GafferImage/ImageReader.cpp:281:31:   required from here
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:66:5: note: boost::intrusive_ptr<T>::intrusive_ptr(T*, bool) [with T = IECore::Data]
     intrusive_ptr( T * p, bool add_ref = true ): px( p )
     ^
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:66:5: note:   no known conversion for argument 1 from 'const boost::intrusive_ptr<const IECore::CompoundObject>' to 'IECore::Data*'
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:62:5: note: boost::intrusive_ptr<T>::intrusive_ptr() [with T = IECore::Data]
     intrusive_ptr() BOOST_NOEXCEPT : px( 0 )
     ^
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:62:5: note:   candidate expects 0 arguments, 1 provided
`

Maybe I'm missing something?How do you cast from a specific type ( like `StringData`) to a generic `Data`

johnhaddon replied:
A cast from specific to generic is automatic, provided the generic is a base class of the specific. But here the issue is that the argument to DefaultColorSpaceFunction is taking CompoundData, but the metadata is a CompoundObject. CompoundObject can't be cast to Data because it doesn't derive from Data.

There are two ways of approaching this. The first is just to make the argument to `DefaultColorSpaceFunction` a `const CompoundObject *` instead of `const CompoundData *`. The other is to redefine the metadata plug so that it carries CompoundData instead of CompoundObject. Although it's a bit more work, I think there's a good argument for the latter - it seems unlikely we'd want to use anything more than data in the metadata, and it seems even less likely that a file format would support more than just data either. @andrewkaufman, I think you were keen on using CompoundObject when you first implemented this - still feel that way?

andrewkaufman replied:
I have no recollection of either being keen on CompoundObject nor why I might have been...

Off the top of my head, could it be because CompoundDataPlug is trying to expose its members as child plugs and we don't have an AtomicCompoundDataPlug ?

johnhaddon replied:
Yeah, looking back through my email, looks like that was it. We only support Data in `ImageReader`, `ImageWriter`, `ImageMetadata` and `ImagePrimitive::blindData()` though, so how about I chuck together a quick PR to add a AtomicCompoundDataPlug and use it for `ImagePlug::metadataPlug()`?

andrewkaufman replied:
Sure

johnhaddon replied:
OK, here we go : https://github.com/GafferHQ/gaffer/pull/2087.

Lucien, once this is merged, could you rebase and confirm that the simple `compoundData->writable()["metadata"] = imagePlug->metadataPlug()->getValue();` approach works? I know this one is dragging on, so give me a shout if you want a hand getting it over the finish line...

lucienfostier replied:
here is what I get

`In file included from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/intrusive_ptr.hpp:16:0,
                 from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/IECore/RefCounted.h:44,
                 from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/IECore/RunTimeTyped.h:39,
                 from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/IECore/Object.h:44,
                 from include/Gaffer/ValuePlug.h:41,
                 from include/Gaffer/StringPlug.h:41,
                 from src/GafferImage/ImageReader.cpp:41:
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp: In instantiation of 'boost::intrusive_ptr<T>& boost::intrusive_ptr<T>::operator=(const boost::intrusive_ptr<U>&) [with U = const IECore::CompoundData; T = IECore::Data]':
src/GafferImage/ImageReader.cpp:270:31:   required from here
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:104:9: error: no matching function for call to 'boost::intrusive_ptr<IECore::Data>::intrusive_ptr(const boost::intrusive_ptr<const IECore::CompoundData>&)'
         this_type(rhs).swap(*this);
         ^
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:104:9: note: candidates are:
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:90:5: note: boost::intrusive_ptr<T>::intrusive_ptr(const boost::intrusive_ptr<T>&) [with T = IECore::Data]
     intrusive_ptr(intrusive_ptr const & rhs): px( rhs.px )
     ^
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:90:5: note:   no known conversion for argument 1 from 'const boost::intrusive_ptr<const IECore::CompoundData>' to 'const boost::intrusive_ptr<IECore::Data>&'
In file included from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/intrusive_ptr.hpp:16:0,
                 from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/IECore/RefCounted.h:44,
                 from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/IECore/RunTimeTyped.h:39,
                 from /home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/IECore/Object.h:44,
                 from include/Gaffer/ValuePlug.h:41,
                 from include/Gaffer/StringPlug.h:41,
                 from src/GafferImage/ImageReader.cpp:41:
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:76:5: note: template<class U> boost::intrusive_ptr<T>::intrusive_ptr(const boost::intrusive_ptr<U>&, typename boost::detail::sp_enable_if_convertible<Y, T>::type)
     intrusive_ptr( intrusive_ptr<U> const & rhs, typename boost::detail::sp_enable_if_convertible<U,T>::type = boost::detail::sp_empty() )
     ^
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:76:5: note:   template argument deduction/substitution failed:
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp: In substitution of 'template<class U> boost::intrusive_ptr<T>::intrusive_ptr(const boost::intrusive_ptr<U>&, typename boost::detail::sp_enable_if_convertible<Y, T>::type) [with U = const IECore::CompoundData]':
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:104:9:   required from 'boost::intrusive_ptr<T>& boost::intrusive_ptr<T>::operator=(const boost::intrusive_ptr<U>&) [with U = const IECore::CompoundData; T = IECore::Data]'
src/GafferImage/ImageReader.cpp:270:31:   required from here
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:76:5: error: no type named 'type' in 'struct boost::detail::sp_enable_if_convertible<const IECore::CompoundData, IECore::Data>'
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp: In instantiation of 'boost::intrusive_ptr<T>& boost::intrusive_ptr<T>::operator=(const boost::intrusive_ptr<U>&) [with U = const IECore::CompoundData; T = IECore::Data]':
src/GafferImage/ImageReader.cpp:270:31:   required from here
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:66:5: note: boost::intrusive_ptr<T>::intrusive_ptr(T*, bool) [with T = IECore::Data]
     intrusive_ptr( T * p, bool add_ref = true ): px( p )
     ^
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:66:5: note:   no known conversion for argument 1 from 'const boost::intrusive_ptr<const IECore::CompoundData>' to 'IECore::Data*'
/home/gafferdev/dev/gaffer/gafferDependencies-0.33.0.0-linux/include/boost/smart_ptr/intrusive_ptr.hpp:62:5: note: boost::intrusive_ptr<T>::intrusive_ptr() [with T = IECore::Data]
     intrusive_ptr() BOOST_NOEXCEPT : px( 0 )
     ^
`

lucienfostier commented on include/GafferImage/ImageReader.h:
```diff
@@ -110,10 +110,18 @@ class ImageReader : public ImageNode
+		typedef boost::function<const std::string ( const IECore::CompoundData * )> DefaultColorSpaceFunction;	
+		static void registerDefaultColorSpace( DefaultColorSpaceFunction f );
+
+		const std::string defaultColorSpace() const;
```
I think this would be useful to be able to query that information for example if you want to inject the original colorspace in the metadata.

johnhaddon replied:
Wouldn't it need to be a plug to make it possible to wire it up via expressions though? I guess maybe it's useful from a testing perspective though...

lucienfostier replied:
Yes I think there is already a plug for that "__intermediateColorSpace", what would you suggest in addition to that?

johnhaddon replied:
Well, like I said, my suggestion would be to keep it all private (like the __intermediateColorSpace plug is)...

lucienfostier replied:
ok sounds good!

lucienfostier commented on src/GafferImage/ImageWriter.cpp:
```diff
@@ -841,6 +874,43 @@ const std::string ImageWriter::currentFileFormat() const
+
+	if( optionsPlug == NULL)
+	{
+		return "";
```
could we? it seems to me that without data type, you cant make a good assumption for the colorspace.
Take for example a tiff file, if you are writing an 8 bits file, using linear is a terrible idea.

johnhaddon replied:
Sure, but for any format that supports multiple data types, we should be providing an option plug, and even for formats which don't support multiple data types, the colourspace function may still have an opinion on what the colourspace should be, based on other knowledge of the format. Not calling the colourspace function at all means it simply has no chance...

lucienfostier replied:
alright that makes sense for jpg for example.

lucienfostier commented on include/GafferImage/ImageReader.h:
```diff
@@ -110,10 +110,18 @@ class ImageReader : public ImageNode
 		static size_t supportedExtensions( std::vector<std::string> &extensions );
 
+		typedef boost::function<const std::string ( const IECore::CompoundData * )> DefaultColorSpaceFunction;	
+		static void registerDefaultColorSpace( DefaultColorSpaceFunction f );
```
ideally we would have the exact same function used in both case, but as you can see from my default implementation that doesnt quite work ( dataType name not consistent between read and write ).Beside I think Gaffer should offer the flexibility to have very different config for reader and writer.Since we also provide an example where the same config is registered for both node, we support also the simpler approach.

johnhaddon replied:
> ideally we would have the exact same function used in both case, but as you can see from my default implementation that doesnt quite work ( dataType name not consistent between read and write ).

I think that might be a bug in OpenImageIOReader's dataType computation though. It seems to report floating point tiffs as uint32 rather than float. I created one using `oiiotool --create 4x4 3 -o float.tif` if you want to try yourself. It'd be good to add some test cases for the dataType plug since it will be so important to this mechanism.

lucienfostier replied:
the dataType in this case comes from the ImageSpec itself, I'm using the `format` data member.

johnhaddon replied:
Are you 100% sure about that? My suspicion is that the `format` data member will be float, but that there is also an "OIIO:BitsPerSample" metadata entry which is triggering the "uint" code path.

lucienfostier replied:
Ok that makes sense, what do you suggest I do to fix this bug? Should I take oiio:BitsPerSample into account only for the dpx file format?

johnhaddon replied:
I think we should probably _start_ with `.format`, and then append `oiio:BitsPerSample` to it if the format is int or uint.

lucienfostier commented:
Hey John,

I already worked on the code to use CompoundObject to fill my CompoundData and everything seems to be working.

I'll rebase and revert to the simpler logic.

I have left a bit of work for updating the test and we should be ready for review/merge.

Cheers

lucienfostier commented:
Hey John,

I think I've addressed all your comments and added the test you advised.
Would you like me to clean up the commit first before you do a review?

Cheers

johnhaddon commented:
Hi Lucien,
That would be great if you could thanks. I've been holding off on looking at this again for a little bit while we focus on getting the next release of the 0.33 series out. When that's done, I'll start looking again at the stuff destined for 0.34...
Cheers...
John

lucienfostier commented:
Hey John,

I've updated the pull request to clean up the commits.
I've include a couple of your commits that are necessary for this to work.
Let me know how you want to merge that part?

Cheers

johnhaddon commented:
Thanks Lucien. Could you rebase on top of the latest master please? That should pick up the metadata changes that Andrew just merged, and resolve the conflict on this branch. The other commit of mine is the GIL management bug - we can keep that on this branch and it'll get merged along with your commits.

lucienfostier commented:
rebase done!

lucienfostier commented:
that's weird, tests run fine locally.

Should we restart them to give it another try?

johnhaddon commented:
> that's weird, tests run fine locally. Should we restart them to give it another try?

Tests fail for me here too, so I don't think restarting will help. It looks like `GafferImageTest.ImageReaderTest.testDataType` is trying to load a file that isn't actually in the repo - perhaps you need to add it? If you do, could you first make sure it's as small as possible please? Also, any chance of making the `chart_*.exr` files smaller too? I'd like to limit the amount of binary data in the repo, and the images for the tests are beginning to add up...

lucienfostier commented:
oh ok, I thought I had changed this to not include unnecessary images in the PR but I'll double check there might be something still calling that.

I'll also take a look at this other image and try to reduce it as much as possible.

Thanks

lucienfostier commented:
Hey John,

I have finally fixed the test and done the commit clean up.
Let me know if you have some comments.

Cheers

lucienfostier commented:
closing in favour of #2121
